### PR TITLE
Started work on the rubber treetap

### PR DESCRIPTION
**Pull request by xana43**
most of the logic is implemented, but the syncing between the gui and the blockstate isn't working yet

### CODE REVIEW

File: src/main/java/io/github/darealturtywurty/turtychemistry/core/network/rubbertreenetworking/ClientBoundUpdateRubberTreeTapPacket.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+    public ClientBoundUpdateRubberTreeTapPacket(final FriendlyByteBuf byteBuf) {
+        this(byteBuf.readBlockPos(), byteBuf.readBoolean());
+    }
+
+    public void encapsulate(final FriendlyByteBuf byteBuf) {
+        byteBuf.writeBlockPos(this.treeTapBlockPos);
+    }
+
+    public void process(Supplier<NetworkEvent.Context> networkContext) {
```
Comment (DaRealTurtyWurty) [inline src/main/java/io/github/darealturtywurty/turtychemistry/core/network/rubbertreenetworking/ClientBoundUpdateRubberTreeTapPacket.java]: Process is a little bit of a weird name. 
```suggestion
    public void handle(Supplier<NetworkEvent.Context> networkContext) {
```

File: src/main/java/io/github/darealturtywurty/turtychemistry/core/network/rubbertreenetworking/ClientBoundUpdateRubberTreeTapPacket.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+        this.isTapping = isTapping;
+    }
+
+    public ClientBoundUpdateRubberTreeTapPacket(final FriendlyByteBuf byteBuf) {
+        this(byteBuf.readBlockPos(), byteBuf.readBoolean());
+    }
+
+    public void encapsulate(final FriendlyByteBuf byteBuf) {
+        byteBuf.writeBlockPos(this.treeTapBlockPos);
+    }
```
Comment (DaRealTurtyWurty) [inline src/main/java/io/github/darealturtywurty/turtychemistry/core/network/rubbertreenetworking/ClientBoundUpdateRubberTreeTapPacket.java]:
```suggestion
    public void encode(final FriendlyByteBuf byteBuf) {
        byteBuf.writeBlockPos(this.treeTapBlockPos);
        byteBuf.writeBoolean(this.isTapping);
    }
```

File: src/main/java/io/github/darealturtywurty/turtychemistry/core/network/rubbertreenetworking/ServerBoundUpdateRubberTreeTapPacket.java [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+        this.isTapping = isTapping;
+    }
+
+    public ServerBoundUpdateRubberTreeTapPacket(final FriendlyByteBuf byteBuf) {
+        this(byteBuf.readBlockPos(), byteBuf.readBoolean());
+    }
+
+    public void encapsulate(final FriendlyByteBuf byteBuf) {
+        byteBuf.writeBlockPos(this.treeTapBlockPos);
+    }
```
Comment (DaRealTurtyWurty) [inline src/main/java/io/github/darealturtywurty/turtychemistry/core/network/rubbertreenetworking/ServerBoundUpdateRubberTreeTapPacket.java]:
```suggestion
    public void encode(final FriendlyByteBuf byteBuf) {
        byteBuf.writeBlockPos(this.treeTapBlockPos);
        byteBuf.writeBoolean(this.isTapping);
    }
```

File: src/main/java/io/github/darealturtywurty/turtychemistry/core/network/rubbertreenetworking/ServerBoundUpdateRubberTreeTapPacket.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+    public void process(Supplier<NetworkEvent.Context> networkContext) {
```
Comment (DaRealTurtyWurty) [inline src/main/java/io/github/darealturtywurty/turtychemistry/core/network/rubbertreenetworking/ServerBoundUpdateRubberTreeTapPacket.java]:
```suggestion
    public void handle(Supplier<NetworkEvent.Context> networkContext) {
```

File: src/main/java/io/github/darealturtywurty/turtychemistry/core/network/PacketHandler.java [hunk 5]
```diff
@@ -16,5 +19,11 @@ public final class PacketHandler extends AbstractInit {
     public static void init() {
         int index = 0;
         TurtyChemistry.LOGGER.info("Registered {} packets!", index);
+        CHANNEL.messageBuilder(ServerBoundUpdateRubberTreeTapPacket.class, index++, NetworkDirection.PLAY_TO_SERVER)
+                .encoder(ServerBoundUpdateRubberTreeTapPacket::encapsulate).decoder(ServerBoundUpdateRubberTreeTapPacket::new)
+                .consumer(ServerBoundUpdateRubberTreeTapPacket::process).add();
+        CHANNEL.messageBuilder(ClientBoundUpdateRubberTreeTapPacket.class, index++, NetworkDirection.PLAY_TO_CLIENT)
+                .encoder(ClientBoundUpdateRubberTreeTapPacket::encapsulate).decoder(ClientBoundUpdateRubberTreeTapPacket::new)
+                .consumer(ClientBoundUpdateRubberTreeTapPacket::process).add();
```
Comment (DaRealTurtyWurty) [inline src/main/java/io/github/darealturtywurty/turtychemistry/core/network/PacketHandler.java]:
```suggestion
        CHANNEL.messageBuilder(ServerBoundUpdateRubberTreeTapPacket.class, index++, NetworkDirection.PLAY_TO_SERVER)
                .encoder(ServerBoundUpdateRubberTreeTapPacket::encode).decoder(ServerBoundUpdateRubberTreeTapPacket::new)
                .consumer(ServerBoundUpdateRubberTreeTapPacket::handle).add();
        CHANNEL.messageBuilder(ClientBoundUpdateRubberTreeTapPacket.class, index++, NetworkDirection.PLAY_TO_CLIENT)
                .encoder(ClientBoundUpdateRubberTreeTapPacket::encode).decoder(ClientBoundUpdateRubberTreeTapPacket::new)
                .consumer(ClientBoundUpdateRubberTreeTapPacket::handle).add();
```

**Review by DaRealTurtyWurty — CHANGES_REQUESTED** [review]

File: src/main/java/io/github/darealturtywurty/turtychemistry/client/screen/RubberTreeTapScreen.java [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+
+    @Override
+    protected void init() {
+        super.init();
+        this.startTappingButtion = addRenderableWidget(
+                new ExtendedButton(this.leftPos, this.topPos, 128, 16, Component.literal("start tapping"),
+                        button -> {
+                            if (minecraft.level.getBlockEntity(BlockPos.ZERO) instanceof RubberTreeTapBlockEntity rubberTreeTapBlockEntity) {
+                                rubberTreeTapBlockEntity.setTapping();
+                            }
```
Comment (DaRealTurtyWurty) [inline src/main/java/io/github/darealturtywurty/turtychemistry/client/screen/RubberTreeTapScreen.java]: What on earth is happening here?

File: src/main/java/io/github/darealturtywurty/turtychemistry/common/block/entity/RubberTreeTapBlockEntity.java [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+    }
+
+    public void setTapping() {
+        TurtyChemistry.LOGGER.info("pressed");
+        if (!this.isTapping) {
+
+            PacketHandler.CHANNEL.sendToServer(new ServerBoundUpdateRubberTreeTapPacket(this.getBlockPos(), true));
+        } else {
+            PacketHandler.CHANNEL.sendToServer(new ServerBoundUpdateRubberTreeTapPacket(this.getBlockPos(), false));
+        }
```
Comment (DaRealTurtyWurty) [inline src/main/java/io/github/darealturtywurty/turtychemistry/common/block/entity/RubberTreeTapBlockEntity.java]:
```suggestion
        PacketHandler.CHANNEL.sendToServer(new ServerBoundUpdateRubberTreeTapPacket(this.getBlockPos(), !this.isTapping)
```

File: src/main/java/io/github/darealturtywurty/turtychemistry/common/block/entity/RubberTreeTapBlockEntity.java [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+        this.inventoryModule = addModule(new InventoryModule(this, 27));
+    }
+
+    @Override
+    public @NotNull ClientboundBlockEntityDataPacket getUpdatePacket() {
+        return ClientboundBlockEntityDataPacket.create(this);
+    }
+
+    public void setTapping() {
+        TurtyChemistry.LOGGER.info("pressed");
```
Comment (DaRealTurtyWurty) [inline src/main/java/io/github/darealturtywurty/turtychemistry/common/block/entity/RubberTreeTapBlockEntity.java]:
```suggestion
        TurtyChemistry.LOGGER.debug("pressed");
```

File: src/main/java/io/github/darealturtywurty/turtychemistry/common/block/treeblocks/rubbertree/RubberTreeBlockStripped.java [hunk 9]
```diff
@@ -31,20 +31,25 @@ public RubberTreeBlockStripped(final Properties properties) {
...
-        final boolean hasRubber = currentState.getValue(HAS_RUBBER);
-        if (currentRubberValue != 0 && hasRubber && !level.isClientSide() /*&& treTapState.getValue(TreeTap.IS_TAPPING)*/) {
-            level.setBlockAndUpdate(pos, currentState.setValue(RUBBER_IN_TREE, currentRubberValue - 2));
+        if (currentRubberValue != 0 /*&& treeTapState.getValue(TreeTap.IS_TAPPING)*/) {
+            level.setBlockAndUpdate(pos, currentState.setValue(RUBBER_IN_TREE, currentRubberValue - 1));
+        } else {
+            level.setBlockAndUpdate(pos, currentState.setValue(HAS_RUBBER, false));
         }
+
+
```
Comment (DaRealTurtyWurty) [inline src/main/java/io/github/darealturtywurty/turtychemistry/common/block/treeblocks/rubbertree/RubberTreeBlockStripped.java]:
```suggestion
```

**Review by DaRealTurtyWurty — CHANGES_REQUESTED** [review]

File: src/main/java/io/github/darealturtywurty/turtychemistry/client/screen/RubberTreeTapScreen.java (hunk 6 above)
Comment (xana43) [reply src/main/java/io/github/darealturtywurty/turtychemistry/client/screen/RubberTreeTapScreen.java]: an attempt to try and get the gui to actually change the blockstate, but as it stands it actually does nothing

File: src/main/java/io/github/darealturtywurty/turtychemistry/common/block/entity/RubberTreeTapBlockEntity.java [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+    }
+
+    @Override
+    public @NotNull ClientboundBlockEntityDataPacket getUpdatePacket() {
+        return ClientboundBlockEntityDataPacket.create(this);
+    }
+
+    public void setTapping() {
+        TurtyChemistry.LOGGER.debug("pressed");
+        PacketHandler.CHANNEL.sendToServer(new ServerBoundUpdateRubberTreeTapPacket(this.getBlockPos(), !this.isTapping)
```
Comment (DaRealTurtyWurty) [inline src/main/java/io/github/darealturtywurty/turtychemistry/common/block/entity/RubberTreeTapBlockEntity.java]:
```suggestion
        PacketHandler.CHANNEL.sendToServer(new ServerBoundUpdateRubberTreeTapPacket(this.getBlockPos(), !this.isTapping);
```

**Review by DaRealTurtyWurty — CHANGES_REQUESTED** [review]

**Comment by Steaf23** [comment]
looks good, maybe the inly thing would be handling the rubber processing from the tap side and having the log just lose rubber passively, but idk how practical that is

**Comment by xana43** [comment]
it doesn't exactly matter which way we do it, because either way it has to random tick to process the rubber, and in both cases the player has to do something in order for it to start checking

**Comment by Steaf23** [comment]
I mean its more of s logic kinda thing, like imo it makes more sense for the tap to actually check the log to see if it can get more latex instead of the other way around it looks like here

**Comment by xana43** [comment]
oh, didn't think of it like that oof, will get on that

**Comment by Steaf23** [comment]
also especially because there will be more logs in the world compared to taps, if only by a little, like lets day you wanna use stripped rubber log in a build, then it would be pointless for them to check for taps. but on the otherhand, a tap is always only connected to a log so just ticking that is cheaper for perfomance in general

File: src/main/java/io/github/darealturtywurty/turtychemistry/client/events/ClientModEvents.java [hunk 11]
```diff
@@ -11,13 +11,20 @@
...
     private ClientModEvents() {
         throw new IllegalStateException("Attempted to construct event bus subscriber!");
     }
 
+
+    @SubscribeEvent
+    public static void clientSetup(FMLClientSetupEvent event) {
+
+    }
+
```
Comment (DaRealTurtyWurty) [inline src/main/java/io/github/darealturtywurty/turtychemistry/client/events/ClientModEvents.java]:
```suggestion
```

File: src/main/java/io/github/darealturtywurty/turtychemistry/client/events/ClientModEvents.java [hunk 12]
```diff
@@ -11,13 +11,20 @@
 import net.minecraftforge.eventbus.api.SubscribeEvent;
 import net.minecraftforge.fml.common.Mod;
 import net.minecraftforge.fml.common.Mod.EventBusSubscriber.Bus;
+import net.minecraftforge.fml.event.lifecycle.FMLClientSetupEvent;
```
Comment (DaRealTurtyWurty) [inline src/main/java/io/github/darealturtywurty/turtychemistry/client/events/ClientModEvents.java]:
```suggestion
```

**Review by DaRealTurtyWurty — CHANGES_REQUESTED** [review]

**Comment by DaRealTurtyWurty** [comment]
Apart from fixing the conflicts and the requested changes, is this PR done?

**Comment by xana43** [comment]
yeah it seems about it, I'm actually trying to commit the requested changes from my phone, and I don't know how to

**Review by DaRealTurtyWurty — APPROVED** [review]

**Pull request merged**